src/sentry/api/endpoints/user_emails.py

```diff
@@ -0,0 +1,240 @@
+from __future__ import absolute_import
+
+import logging
+
+from django.db import IntegrityError, transaction
+from django.db.models import Q
+from rest_framework import serializers, status
+from rest_framework.response import Response
+
+from sentry import newsletter
+from sentry.api.bases.user import UserEndpoint
+from sentry.api.decorators import sudo_required
+from sentry.api.serializers import serialize
+from sentry.models import User, UserEmail, UserOption
+
+logger = logging.getLogger('sentry.accounts')
+
+
+InvalidEmailResponse = Response({'detail': 'Invalid email'}, status=status.HTTP_400_BAD_REQUEST)
+
+
+class InvalidEmailError(Exception):
+    pass
+
+
+class DuplicateEmailError(Exception):
+    pass
+
+
+class EmailSerializer(serializers.Serializer):
+    email = serializers.EmailField(required=False)
+
+
+def get_email(request):
+    """
+    Returns None if invalid email else returns a normalized email (lowercase + whitespace strip)
+    """
+
+    serializer = EmailSerializer(data=request.DATA)
+
+    # Bad email
+    if not serializer.is_valid():
+        return None
+
+    return serializer.object['email'].lower().strip()
+
+
+def add_email(email, user):
+    """
+    Adds an email to user account
+
+    Can be either primary or secondary
+    """
+
+    # Bad email
+    if email is None:
+        raise InvalidEmailError
+
+    # check if this email already exists for user
+    if email and UserEmail.objects.filter(
+        user=user, email__iexact=email
+    ).exists():
+        raise DuplicateEmailError
+
+    try:
+        with transaction.atomic():
+            new_email = UserEmail.objects.create(user=user, email=email)
+    except IntegrityError:
+        raise DuplicateEmailError
+    else:
+        new_email.set_hash()
+        new_email.save()
+        user.send_confirm_email_singular(new_email)
+
+        # Update newsletter subscription and mark as unverified
+        newsletter.update_subscription(user=user,
```

Comment:
This is from `frontend/accounts.py` (which also throws an error), but doesn't work for me locally running w/ gs, I get this error:

`TypeError: update_subscription() takes at least 3 arguments (3 given)`

Reply:
according to stack overflow, this means you added an option argument so you hit the count of 3, but there was a required argument you missed. (self counts as the first arg, if you hadn't noticed)

Reply:
Definitely not going to merge until I understand what's going on

Reply:
So it looks like `getsentry` overrides the `Newsletter` class and adds a `ConsentRequiredNewsletter` which has changed the function signature. It added another argument which isn't in the base class.

So no matter what, it'll be broken in both places.

@mikeclarke not sure if this was intentional or an oversight, but this seems less than idea considering it won't work at all.

This is also only the backend for `dev`, and `prod` uses `getsentry.newsletter.LocalNewsletter`.

Reply:
So I wouldn't worry about this error at this time. It's a problem with the newsletter backend that isn't used in production yet. Let's just make a note of it so @mikeclarke can follow up and either change the signature all around, or fix it another way.